spark_connect_labextension/sparkconnectserver/cluster.py

```diff
@@ -0,0 +1,227 @@
+        if retcode != 0:
+            raise Exception("Cannot start Spark Connect server")
+
+    def stop(self):
+        """
+        Stop the running Spark Connect server
+        """
+        print("Stopping Spark Connect server...")
+        env_variables = self.get_envs()
+        run_script = f"sh $SPARK_HOME/sbin/stop-connect-server.sh"
+        if self.pre_script:
+            run_script = self.pre_script + ' && ' + run_script
+
+        retcode = subprocess.Popen(run_script, shell=True, env=env_variables).wait()
+        if retcode != 0:
+            raise Exception("Cannot stop Spark Connect server")
+        self.started = False
+
+    def get_log(self) -> str:
+        """
+        Retrieve the Spark Connect server logs
+
+        :returns: log string
+        """
+        logfile = self.get_logfile_path()
+        if not logfile:
+            return None
+
+        with open(logfile, 'r') as f:
+            return f.read()
+
+    def get_logfile_path(self) -> str:
+        """
+        Get the log file path
+
+        :returns: log file path
+        """
+        files = glob.glob(self.tmpdir.name + '/*.out')
+        if len(files) == 0:
+            return None
+        logfile = files[-1]
+        return logfile
+
+    def get_status(self) -> ClusterStatus:
+        """
+        Get the status of the connection
+
+        :returns: ClusterStatus
+        """
+        if not self.is_connect_server_running():
+            return ClusterStatus.STOPPED
+        if not self.is_server_ready():
+            return ClusterStatus.PROVISIONING
+        return ClusterStatus.READY
+
+    def get_port(self) -> int:
+        """
+        Get the Spark Connect server port
+
+        :returns: port number
+        """
+        return SPARK_CONNECT_PORT
+
+    def get_options(self) -> dict:
+        """
+        Get the Spark options for running the Connect server
+
+        :returns: dict of options
+        """
+        return self.options
+
+    def is_connect_server_running(self) -> bool:
+        """
+        Determine if the Spark Connect server process is running
+
+        :returns: boolean indicating if the server is running
+        """
+        env_variables = self.get_envs()
+        run_script = f"sh $SPARK_HOME/sbin/spark-daemon.sh status org.apache.spark.sql.connect.service.SparkConnectServer 1"
+        if self.pre_script:
+            run_script = self.pre_script + ' && ' + run_script
+
+        retcode = subprocess.Popen(run_script, shell=True, env=env_variables).wait()
+        return retcode == 0
+
+    def is_server_ready(self) -> bool:
+        """
+        Determine if the Spark Connect server is ready and exposing the gRPC socket through the specified port.
+
+        :returns: boolean indicating if the server is ready
+        """
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        result = sock.connect_ex(('127.0.0.1', self.get_port()))
+        sock.close()
```

Comment:
This can all throw an exception, right? Better to protect with try except, and then you reraise a new exception from the one you intercepted.
```
try:
   ...
except SomeException as e:
    raise Exception('Cannot determine if the Spark Connect server is ready') from e
```

Reply:
Why do we need this?

Reply:
This exception wrapping allows you to raise an exception with a meaningful message for your application, and still preserve the original cause of the exception. Allowing low-level exceptions (in this case, socket connection problems) to be propagated freely makes it more difficult to debug. Having an extra explanatory message (more high-level) helps debugging.

Reply:
Okay got it, thanks!